go/pkg/pass1/set-path.go

```diff
@@ -287,556 +222,497 @@ Partition:
-				}
-			}
-		}
+// Pathrestrictions in loops with > 1 zone cluster affect ACLs.
+func restrictIsInLoopWithSeveralZoneClusters(restrict *pathRestriction) bool {
 
-		if len(intfZoneClusters) > 0 {
-			if len(intfZoneClusters) != len(intfZones) {
-				continue Restrict
-			} else {
-				referenceZoneCluster := intfZoneClusters[0]
-				for _, zoneClusterElem := range intfZoneClusters {
-					for j, zoneElem := range zoneClusterElem {
-						if referenceZoneCluster[j] != zoneElem {
-							continue Restrict
-						}
-					}
-				}
-			}
-		}
-		// Pathrestrictions in loops with > 1 zone cluster have an effect.
-		intf := elements[0]
-		intfLoop := getIntfLoop(intf)
-		intfZone := intf.zone
-		zoneCluster := intfZone.zoneCluster
-		if len(zoneCluster) == 0 {
-			zoneCluster = append(zoneCluster, intfZone)
-		}
+	referenceLoop := getIntfLoop(restrict.elements[0])
+	referenceCluster := restrict.elements[0].zone.zoneCluster
+	if len(referenceCluster) == 0 {
+		referenceCluster = append(referenceCluster, restrict.elements[0].zone)
+	}
 
-		// Process every zone in zone cluster...
-		for _, zone1 := range zoneCluster {
-			for _, intf1 := range zone1.interfaces {
-				r := intf1.router
+	for _, z := range referenceCluster {
+		for _, zoneIntf := range z.interfaces {
+			r := zoneIntf.router
 
-				// ...examine its neighbour zones:
-				for _, intf2 := range r.interfaces {
-					zone2 := intf2.zone
-					if zone2 == intfZone {
-						continue
-					}
-					if zone2.zoneCluster != nil {
-						cluster2 := zone2.zoneCluster
+			for _, intf2 := range r.interfaces {
+				zone2 := intf2.zone
+				if zone2 == z {
+					continue
+				}
+				if zone2.zoneCluster != nil {
 
-						if areEqualZoneClusters(zoneCluster, cluster2) {
-							continue
-						}
+					if areEqualZoneClusters(referenceCluster, zone2.zoneCluster) {
+						continue
 					}
-					if zone2.getLoop() != nil {
-						loop2 := zone2.getLoop()
-						if intfLoop == loop2 {
-							// Found other zone cluster in same loop.
-							continue Restrict
-						}
+				}
+				if zone2.getLoop() != nil {
+					if referenceLoop == zone2.getLoop() {
+						return true
 					}
 				}
 			}
 		}
-
-		warnMsg("Useless %s.\n All interfaces are unmanaged and located inside the same security zone", restrict.name)
-
-		// Clear interfaces of useless pathrestriction.
-		restrict.elements = nil
 	}
+	return false
+}
 
-	// Collect all effective pathrestrictions.
-	for _, restrict := range pathrestrictions {
-		if restrict.elements != nil {
-			effectivePathrestrictions = append(effectivePathrestrictions, restrict)
+func areEqualZoneClusters(cluster1 []*zone, cluster2 []*zone) bool {
+	if len(cluster1) != len(cluster2) {
+		return false
+	}
+	sort.Sort(zoneNameSorter(cluster1))
+	sort.Sort(zoneNameSorter(cluster2))
+	for index := 0; index < len(cluster1); index++ {
+		if cluster1[index] != cluster2[index] {
+			return false
 		}
 	}
-	sort.Sort(NameSorter(effectivePathrestrictions))
+	return true
+}
```

Comment:
Das scheint mit zu aufwändig im Vergleich zu "$cluster2 eq $zone_cluster" in Perl.

Reply:
:facepalm: Danke!